File: android/src/main/kotlin/com/crefter/yuchengplugin/yucheng_ble/YuchengSleepDataConverter.kt
package com.crefter.yuchengplugin.yucheng_ble

import YuchengSleepData
import YuchengSleepDataDetail
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.yucheng.ycbtsdk.Constants

private class SleepDataBean(
    val deepSleepCount: Int,// deepSleepCount
    val lightSleepCount: Int, // lightSleepCount
    val startTime: Long,// startSleepTime
    val endTime: Long?, // endSleepTime
    val deepSleepTotal: Int, // deepSleepTotal
    val lightSleepTotal: Int, // lightSleepTotal
    @SerializedName(
        value =
            "rapidEyeMovementTotal", alternate = ["remTimes"]
    )
    val rapidEyeMovementTotal: Int, // eyeMovementTotal
    val wakeCount: Int,// number of times awake
    val wakeDuration: Int,// length of time awake
    val sleepData: List<SleepData> = ArrayList(), // sleep data
)

private class SleepData(
    @SerializedName(
        value =
            "sleepStartTime", alternate = ["stime"]
    )
    val sleepStartTime: Long, // start time

    @SerializedName(
        value =
            "sleepLen", alternate = ["sleepLong"]
    )
    val sleepLen: Int, // sleep length
    var sleepType: Int, // deepSleepLightSleep flag, flag type is SleepType below)
)

class YuchengSleepDataConverter(private val gson: Gson) {
    fun convert(sleepDataBean: Any?): YuchengSleepData {
        val converted = gson.fromJson(sleepDataBean.toString(), SleepDataBean::class.java)
        val deepSleepCount = converted.deepSleepCount.toLong()
        val isOldFormat = deepSleepCount.toInt() != 0xFFFF
        val deepSleepInSeconds = formatSleepDuration(converted.deepSleepTotal, isOldFormat)
        val remSleepInSeconds = formatSleepDuration(converted.rapidEyeMovementTotal, isOldFormat)
        val lightSleepInSeconds = formatSleepDuration(converted.lightSleepTotal, isOldFormat)
        val awakeSleepInSeconds = formatSleepDuration(converted.wakeDuration, isOldFormat)

        val details = converted.sleepData.map {
                detail ->
            val sleepType = when(detail.sleepType)
            {
                Constants.SleepType.deepSleep -> YuchengSleepType.DEEP
                Constants.SleepType.lightSleep -> YuchengSleepType.LIGHT
                Constants.SleepType.rem -> YuchengSleepType.REM
                Constants.SleepType.awake -> YuchengSleepType.AWAKE
                else -> YuchengSleepType.UNKNOWN
            }
            YuchengSleepDataDetail(
                startTimeStamp = detail.sleepStartTime,
                duration = detail.sleepLen.toLong(),
                type = sleepType
            )
        }
        return YuchengSleepData(
            startTimeStamp = converted.startTime,
            endTimeStamp = converted.endTime ?: 0,
            deepCount = deepSleepCount,
            lightCount = converted.lightSleepCount.toLong(),
            awakeCount = converted.wakeCount.toLong(),
            deepInSeconds = deepSleepInSeconds.toLong(),
            remInSeconds = remSleepInSeconds.toLong(),
            lightInSeconds = lightSleepInSeconds.toLong(),
            awakeInSeconds = awakeSleepInSeconds.toLong(),
            details = details,
        )
    }

    private fun formatSleepDuration(duration: Int, isOldFormat: Boolean): Int {
        return duration * (if (isOldFormat) 60 else 1)
    }
}

File: ios/Classes/YuchengSleepDataConverter.swift
//
//  YuchengSleepDataConverter.swift
//  Pods
//
//  Created by Maxim Zarechnev on 10.04.2025.
//

import YCProductSDK

final class YuchengSleepDataConverter: Sendable {
    func convert(sleepDataFromDevice: YCHealthDataSleep) -> YuchengSleepData {
        var sleepDetails: [YuchengSleepDataDetail] = []
        var awakeCount = 0;
        var awakeSeconds = 0;
        for detail in sleepDataFromDevice.sleepDetailDatas {
            var type = YuchengSleepType.unknown
            let detailType = detail.sleepType
            if (detailType == YCHealthDataSleepType.awake) {
                type = .awake
                awakeCount += 1
                awakeSeconds += Int(detail.duration)
            } else if (detailType == YCHealthDataSleepType.deepSleep) {
                type = .deep
            } else if (detailType == YCHealthDataSleepType.lightSleep) {
                type = .light
            } else if (detailType == YCHealthDataSleepType.unknow) {
                type = .unknown
            } else if (detailType == YCHealthDataSleepType.rem) {
                type = .rem
            }
            sleepDetails.append(YuchengSleepDataDetail(startTimeStamp: Int64(detail.startTimeStamp), duration: Int64(detail.duration), type: type))
        }
        let isOldFormat = sleepDataFromDevice.deepSleepCount != 0xFFFF;
        
        let deepSeconds = isOldFormat ? sleepDataFromDevice.deepSleepMinutes * 60 : sleepDataFromDevice.deepSleepSeconds
        let lightSeconds = isOldFormat ? sleepDataFromDevice.lightSleepMinutes * 60 : sleepDataFromDevice.lightSleepSeconds
        let remSeconds = isOldFormat ? sleepDataFromDevice.remSleepMinutes * 60 : sleepDataFromDevice.remSleepSeconds
        
        let event = YuchengSleepData(startTimeStamp: Int64(sleepDataFromDevice.startTimeStamp).toMilliseconds(), endTimeStamp: Int64(sleepDataFromDevice.endTimeStamp).toMilliseconds(), deepCount: Int64(sleepDataFromDevice.deepSleepCount), lightCount: Int64(sleepDataFromDevice.lightSleepCount),  awakeCount: Int64(awakeCount), deepInSeconds: Int64(deepSeconds), remInSeconds: Int64(remSeconds), lightInSeconds: Int64(lightSeconds), awakeInSeconds: Int64(awakeSeconds), details: sleepDetails )
        return event
    }
}
